sled-agent/src/sim/sled_agent.rs

```diff
@@ -132,4 +134,87 @@ impl SledAgent {
     ) -> Arc<CrucibleData> {
         self.storage.lock().await.get_dataset(zpool_id, dataset_id).await
     }
+
+    /// Get contents of an instance's serial console.
+    pub async fn instance_serial_console_data(
+        &self,
+        instance_id: Uuid,
+        byte_offset: Option<isize>,
+        max_bytes: Option<usize>,
+    ) -> Result<InstanceSerialConsoleData, String> {
+        if !self.instances.sim_contains(&instance_id).await {
+            return Err(format!("No such instance {}", instance_id));
+        }
+
+        // TODO: if instance state isn't running {
+        //  return Ok(InstanceSerialConsoleData { data: vec![], last_byte_offset: 0 });
+        // }
+
+        let gerunds = [
+            "Loading",
+            "Reloading",
+            "Advancing",
+            "Reticulating",
+            "Defeating",
+            "Spoiling",
+            "Cooking",
+            "Destroying",
+            "Resenting",
+            "Introducing",
+            "Reiterating",
+            "Blasting",
+            "Tolling",
+            "Delivering",
+            "Engendering",
+            "Establishing",
+        ];
+        let nouns = [
+            "canon",
+            "browsers",
+            "meta",
+            "splines",
+            "villains",
+            "plot",
+            "books",
+            "evidence",
+            "decisions",
+            "chaos",
+            "points",
+            "processors",
+            "bells",
+            "value",
+            "gender",
+            "shots",
+        ];
+        let mut entropy = instance_id.as_u128();
+        let mut buf = format!(
+            "This is simulated serial console output for {}.\n",
+            instance_id
+        );
+        while entropy != 0 {
+            let gerund = gerunds[entropy as usize % gerunds.len()];
+            entropy /= gerunds.len() as u128;
+            let noun = nouns[entropy as usize % nouns.len()];
+            entropy /= nouns.len() as u128;
+            buf += &format!(
+                "{} {}... {}[\x1b[92m 0K \x1b[m]\n",
+                gerund,
+                noun,
+                " ".repeat(40 - gerund.len() - noun.len())
+            );
+        }
+        buf += "\x1b[2J\x1b[HOS/478 (localhorse) (ttyl)\n\nlocalhorse login: ";
```

Comment:
The legend of localhorse continues @jordanhendricks  😅️

Reply:
(OS/478 was kind of my take on a fictionalized far-future OS/2 successor, the `0x1DE`-th in the franchise)